[JIT] Trace uses of torchbind classes as module attributes

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32938 [WIP][quantization] Use torchbind for Linear PackedParams
* #32977 [JIT] Resolve custom classes in source importer
* #32948 [c10] Allow taking a std::tuple as arg
* #32878 [JIT] Fix python pickle serialization for torchbind
* #32854 [JIT] Make IRParser use op schema
* **#32833 [JIT] Trace uses of torchbind classes as module attributes**

Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

Commit:
[JIT] Trace uses of torchbind classes as module attributes

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9206eec9:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9206eec9483bbc81376617be6f62faaebfaff994).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4382381) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93371780) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93371780))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
Auto-merging torch/csrc/jit/script/init.cpp 
CONFLICT (content): Merge conflict in torch/csrc/jit/script/init.cpp 
Removing tools/autograd/templates/python_variable_methods_dispatch.h 
Removing tools/autograd/templates/python_torch_functions_dispatch.h 
Removing tools/autograd/templates/python_nn_functions_dispatch.h 
Auto-merging test/test_jit.py 
Removing aten/src/THNN/generic/BCECriterion.c 
Removing aten/src/THCUNN/generic/BCECriterion.cu 
Removing aten/src/THCUNN/BCECriterion.cu 
Removing aten/src/ATen/detail/ScalarTypeConversions.h 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

suo commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -148,6 +149,27 @@ Value* TracingState::getValue(const IValue& var) {
+    if (var.isObject()) {
+      auto obj = script::Object(var.toObject());
+      auto qualname = obj.type()->name();
+      if (qualname) {
```
when is there not a qualname? Should this be a hard error?

suo commented on torch/jit/__init__.py:
```diff
@@ -1870,6 +1870,9 @@ def check_unique(param):
                 tmp_module._buffers[name] = buf
                 check_unique(buf)
+        for name, val in orig.__dict__.items():
+            if torch._C._jit_is_traceable_ivalue(val) and name not in orig._parameters and name not in orig._buffers:
```
This condition is a little opaque without context, can we:
1. add a comment explaining what the rule here is, and 2
2. Add `orig._modules` so its clear that we're saying "python attributes that not part of the `nn.Module` state"

suo replied:
we should also check_unique here I guess

suo commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -408,6 +437,11 @@ void TracingState::setValue(const IValue& v, Value* value) {
       setValue(elements[i], unpack_node->outputs()[i]);
     }
+  } else if (
+      v.isObject() && v.toObject()->type()->name() &&
```
seems that in 3 places we're doing a multi-step unwrap that boils down to `bool isCustomClass(const IValue&)`—can we turn it into a utility function defined in `custom_class.h`

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -1330,6 +1330,9 @@ void initJitScriptBindings(PyObject* module) {
         },
         py::arg("proto_string")
       );
+  m.def("_jit_is_traceable_ivalue", [](const py::object& obj) {
```
The name of this function is probably too generic—for example, it returns false for some traceable ivalues (tensor, etc.). Possibly someone will use it expecting it to do the right thing and be sad. Can we change either the name or the impl? If we change the impl, can we use it in the tracer so that there is some guarantee it remains correct?

suo reviewed (approved):
So stuff like `setattr` doesn't work for parameters today?

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]

Commit:
Update on "[JIT] Trace uses of torchbind classes as module attributes"


Differential Revision: [D19645714](https://our.internmc.facebook.com/intern/diff/D19645714)

[ghstack-poisoned]
